File: 2024/11/main.py
#!/usr/bin/env python3

import sys

csMem = {}
def countStone(stone, time):
    if (stone,time) in csMem:
        return csMem[(stone,time)]

    stoneString = str(stone)
    stoneStringLength = len(stoneString)
    if time == 0:
        return 1
    elif (stone == 0):
        ans = countStone(1, time-1)
    elif len(stoneString) % 2 == 0:
        stone1 = int(stoneString[:stoneStringLength//2])
        stone2 = int(stoneString[stoneStringLength//2:])
        ans = countStone(stone1,time-1) + countStone(stone2,time-1)
    else:
        ans = countStone(2024*stone, time-1)

    csMem[(stone,time)] = ans
    return ans

def doPart1():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    data = [int(x) for x in open(filename,"r").read().strip().split()]

    ans = 0
    t = 25
    for d in data:
        curAnswer = countStone(d,t)
        ans += curAnswer
    print(ans)

def doPart2():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    data = [int(x) for x in open(filename,"r").read().strip().split()]

    ans = 0
    t = 75
    for d in data:
        ans += countStone(d,t)
    print(ans)

def main():
    doPart1()
    doPart2()

if __name__ == "__main__":
    main()

File: 2024/11/main.cpp

#include "../helpers/extras.hh"
#include <unordered_map>

unsigned long countStone(
    const unsigned long stone,
    const int time,
    std::unordered_map<unsigned long, unsigned long>& csMEM
)
{
    if (csMEM.find(1000*stone + time) != csMEM.end()) {
        return csMEM[1000*stone + time];
    }
    std::string stoneString = std::to_string(stone);
    const size_t stoneStringLength = stoneString.length();
    unsigned long ans = 0;
    if (time == 0) {
        return 1;
    } else if (stone == 0) {
        ans = countStone(1, time-1, csMEM);
    } else if (stoneStringLength % 2 == 0) {
        unsigned long stone1 = std::atoi(stoneString.substr(0, stoneStringLength/2).c_str());
        unsigned long stone2 = std::atoi(stoneString.substr(stoneStringLength/2,stoneStringLength).c_str());
        ans = countStone(stone1, time-1, csMEM) + countStone(stone2, time-1, csMEM);
    } else {
        ans = countStone(2024*stone, time-1, csMEM);
    }

    csMEM[1000*stone+time] = ans;

    return ans;
}

void doPart1(const char* filename)
{
    std::vector<unsigned long> data;
    std::ifstream file(filename);
    std::string line;
    getline(file, line);
    file.close();
    stripString(line);
    std::vector<std::string> sLine = splitString(line, " ");
    for (std::string val : sLine) {
        data.push_back(std::atol(val.c_str()));
    }

    std::unordered_map<unsigned long, unsigned long> csMEM;
    unsigned long ans = 0;
    unsigned long curAnswer = 0;
    for (unsigned long v : data) {
        curAnswer = countStone(v, 25, csMEM);
        ans += curAnswer;
    }
    printf("%ld\n", ans);
}

void doPart2(const char* filename)
{
    std::vector<unsigned int> data;
    std::ifstream file(filename);
    std::string line;
    getline(file, line);
    file.close();
    stripString(line);
    std::vector<std::string> sLine = splitString(line, " ");
    for (std::string val : sLine) {
        data.push_back(std::atol(val.c_str()));
    }

    std::unordered_map<unsigned long, unsigned long> csMEM;
    unsigned long ans = 0;
    for (unsigned long v : data) {
        ans += countStone(v, 75, csMEM);
    }
    printf("%ld\n", ans);
}

int main(int narg, char* args[])
{
    std::string filename = "example.dat";
    if (narg > 1) filename = std::string(args[1]);
    doPart1(filename.c_str());
    doPart2(filename.c_str());
    return 0;
}
